src/components/snippetpanel.ts

```diff
@@ -0,0 +1,230 @@
+
+        const mathSymbolPromises: Promise<any>[] = []
+        for (const category in snippets.mathSymbols) {
+            for (let i = 0; i < snippets.mathSymbols[category].length; i++) {
+                const symbol = snippets.mathSymbols[category][i]
+                if (symbol.svg === undefined) {
+                    mathSymbolPromises.push(
+                        new Promise((resolve, reject) => {
+                            this.mathJax
+                                .typeset({
+                                    math: symbol.source,
+                                    format: 'TeX',
+                                    svgNode: true
+                                })
+                                .then(
+                                    (data: {
+                                        height: string;
+                                        speakText: string;
+                                        style: string;
+                                        svgNode: SVGSVGElement;
+                                        width: string;
+                                    }) => {
+                                        let svg = data.svgNode.outerHTML
+                                        svg = svg.replace(
+                                            /<title([^>]*)>(.*)<\/title>/,
+                                            `<title$1>${symbol.name.toLocaleUpperCase()}.${
+                                                symbol.keywords ? ' Keywords: ' + symbol.keywords : ''
+                                            }</title>`
+                                        )
+                                        if (symbol.shrink) {
+                                            svg = svg.replace(/^<svg/, '<svg class="shrink"')
+                                        }
+                                        symbol.svg = svg
+                                        resolve()
+                                    }
+                                )
+                                .catch(reject)
+                        })
+                    )
+                }
+            }
+        }
+        Promise.all(mathSymbolPromises).finally(() => {
+            if (mathSymbolPromises.length > 0) {
+                writeFileSync(snipetsFile, JSON.stringify(snippets, undefined, 4))
+                vscode.window.showInformationMessage(
+                    `LaTeX-Workshop: ${mathSymbolPromises.length} symbols rendered and cached`
+                )
+            }
+            for (const category in snippets.mathSymbols) {
+                for (let i = 0; i < snippets.mathSymbols[category].length; i++) {
+                    const symbol = snippets.mathSymbols[category][i]
+                    symbol.category = category
+                    if (symbol.keywords === undefined) {
+                        symbol.keywords = ''
+                    }
+                    // @ts-ignore
```

Comment:
TSC thinks that `symbol.keywords` could be undefined, so it produces the message
```
Argument of type '{ name: string; keywords?: string | undefined; category?: string | undefined; source: string; snippet: string; svg?: string | undefined; shrink?: boolean | undefined; }' is not assignable to parameter of type 'IMathSymbol'.
  Types of property 'category' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.
```
However due to 187-188 `symbol.keywords` will always be a string, and so tsc is incorrect.

Reply:
Don't use `@ts-ignore`. Just fix it.

Reply:
I cannot see why you do not use `IMathSymbol` as follows:
```typescript
const snippets: {
            mathSymbols: {
                [category: string]: IMathSymbol[];
            };
        } = JSON.parse(readFileSync(snipetsFile, { encoding: 'utf8' }))
```

How about this patch? We should use `type` instead of `interface` if there is not a specific reason.

```diff
diff --git a/src/components/snippetpanel.ts b/src/components/snippetpanel.ts
index 55d0af2..52ad688 100644
--- a/src/components/snippetpanel.ts
+++ b/src/components/snippetpanel.ts
@@ -5,13 +5,14 @@ import * as fs from 'fs'
 
 import { Extension } from '../main'
 
-interface IMathSymbol {
+type IMathSymbol = {
     name: string
     keywords?: string
     source: string
     snippet: string
-    category: string
+    category?: string
     svg?: string
+    shrink?: boolean
 }
 
 export class SnippetPanel {
@@ -120,15 +121,7 @@ export class SnippetPanel {
         const snipetsFile = path.join(this.extension.extensionRoot, 'resources', 'snippetpanel', 'snippetpanel.json')
         const snippets: {
             mathSymbols: {
-                [category: string]: {
-                    name: string;
-                    keywords?: string;
-                    category?: string;
-                    source: string;
-                    snippet: string;
-                    svg?: string;
-                    shrink?: boolean;
-                }[];
+                [category: string]: IMathSymbol[];
             };
         } = JSON.parse(readFileSync(snipetsFile, { encoding: 'utf8' }))
 
@@ -187,7 +180,6 @@ export class SnippetPanel {
                     if (symbol.keywords === undefined) {
                         symbol.keywords = ''
                     }
-                    // @ts-ignore
                     this.mathSymbols.push(symbol)
                 }
             }
```

Reply:
Done in b6f923ea51b9f786726df114c3e809f0071e9763.